aioshelly/coap.py

```diff
@@ -69,6 +86,74 @@ def socket_init():
     return sock
 
 
+def verify_l2socket_creation_permission():
+    """Create a socket using the scapy configured l2socket.
+
+    Try to create the socket
+    to see if we have permissions
+    since sniff will do it another
+    thread so we will not be able to capture
+    any permission or bind errors.
+    """
+    conf.L2socket()
+
+
+class MulticastQuerier:
+    """Multicast querier management."""
+
+    def __init__(self):
+        """Initialize multicast querier thread."""
+        try:
+            verify_l2socket_creation_permission()
+        except (Scapy_Exception, OSError) as ex:
+            if os.geteuid() == 0:
+                _LOGGER.error("Cannot watch for multicast query packets: %s", ex)
+            else:
+                _LOGGER.debug(
+                    "Cannot watch for multicast query packets without root or CAP_NET_RAW: %s",
+                    ex,
+                )
+            return
+        _LOGGER.debug("Multicast querier thread started")
+        self.stop_thread = False
+        self.thread = threading.Thread(target=self.run)
+        self.thread.start()
+
+    def run(self):
+        """Start sniffing for multicast query."""
+        filter_igmp_query = f"igmp and igmp[0] == {int(MULTICAST_QUERY_IGMPTYPE)} and (igmp[4:4] == {int(ipaddress.IPv4Address(MAIN_MULTICAST_IP))} or igmp[4:4] == 0)"
+        pkt_snd = IP(dst=MULTICAST_QUERY_GRP) / IGMP(
+            type=MULTICAST_QUERY_IGMPTYPE,
+            mrcode=100,
+            gaddr=MAIN_MULTICAST_IP,
+        )
+        while not self.stop_thread:
+            result = sniff(
+                filter=filter_igmp_query,
+                store=1,
+                count=1,
+                timeout=MULTICAST_QUERY_TIMEOUT,
+            )
+            _LOGGER.debug("Multicast query sniff result: %s", result)
+            if not result:
+                _LOGGER.info(
+                    "Multicast query not received in %s seconds",
+                    MULTICAST_QUERY_TIMEOUT,
+                )
+                for eth_iface in get_network_objects("iface"):
+                    send(pkt_snd, iface=eth_iface, verbose=False)
+                    _LOGGER.info("Multicast query sent for %s interface", eth_iface)
+            else:
+                _LOGGER.info(
+                    "Multicast query received from network, no action required"
```

Comment:
How will this deal with things like the Shelly Motion? It's a battery device that is available on the network. Will it cause the device to wake up the whole time and acknowledge membership?

Reply:
Multicast query are usually sent by the router, so are always present on a well working IGMP network.
In any case we don't create any traffic to those devices, as they seems not to listen to queries.

Reply:
@chemelli74 I think we should have a flag in `initialize` (default can be true) so that we can disable querier without releasing a new version if it makes problems.